pkg/system/system.go

```diff
@@ -0,0 +1,225 @@
+					return *d.RootMountPoint
+				}
+				return ""
+			}
+
+			CustomMountPointGetter := func() string {
+				if d.CustomMountPoint != nil {
+					return *d.CustomMountPoint
+				}
+				return ""
+			}
+
+			i := daemonInfo{ID: d.ID,
+				SnapshotID:       d.SnapshotID,
+				Pid:              d.Pid,
+				ImageID:          d.ImageID,
+				RootMountPoint:   RootMountPointGetter(),
+				CustomMountPoint: CustomMountPointGetter(),
+				SnapshotDir:      d.SnapshotDir}
+
+			info = append(info, i)
+		}
+
+		w.WriteHeader(http.StatusOK)
+		w.Header().Set("Content-Type", "application/json")
+
+		respBody, err := json.Marshal(&info)
+		if err != nil {
+			panic("marshal")
+		}
+
+		w.Write(respBody)
+	}
+}
+
+// TODO: Implement me!
+func (sc *Controller) getDaemonRecords() func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		daemons := sc.manager.ListDaemons()
+		log.L.Infof("list daemons %v", daemons)
+	}
+}
+
+// POST /api/v1/nydusd/upgrade
+// body: {"nydusd_path": "/path/to/new/nydusd", "version": "v2.2.1", "policy": "rolling"}
+// Possible policy: rolling, immediate
+// Live upgrade procedure:
+//  1. Check if new version of nydusd executive is existed.
+//  2. Validate its version matching `version` in this request.
+//  3. Upgrade one nydusd:
+//     a. Lock the whole manager daemons cache, no daemon can be inserted of deleted from manager
+//     b. Start a new nydusd with `--upgrade` flag, wait until it reaches INTI state
+//     c. Validate the new nydusd's version returned by API /daemon
+//     d. Send resources like FD and daemon running states to the new nydusd by API /takeover
+//     e. Wait until new nydusd reaches state READY
+//     f. Command the old nydusd to exit
+//     g. Send API /start to the new nydusd making it take over the whole file system service
+//
+// 4. Upgrade next nydusd like step 3.
+// 5. If upgrading a certain nydusd fails, abort!
+// 6. Delete the old nydusd executive
+func (sc *Controller) upgradeDaemons() func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		daemons := sc.manager.ListDaemons()
+
+		var c upgradeRequest
+
+		err := json.NewDecoder(r.Body).Decode(&c)
+		if err != nil {
+			log.L.Errorf("decode error, %v", r)
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		// TODO: Keep the nydusd executive path in Daemon state and persis it since nydusd
+		// can run on both versions.
+		// Create a dedicated directory storing nydusd of various versions?
+
+		// TODO: Let struct `Daemon` have a method to start a process rather than forking
+		// process in daemons manager
+		// TODO: daemon client has a method to query daemon version and information.
+		for _, d := range daemons {
+			err := upgradeNydusDaemon(d, c)
+			if err != nil {
+				log.L.Errorf("Upgrade daemon %s failed", d.ID)
+				http.Error(w, err.Error(), http.StatusInternalServerError)
```

Comment:
Better to define the error structure like the one below, instead of responding with a plain error message:

```
Status Code: 500

Response Body:
{
    code: "ERR_UPGRADE_FAILED",
    message: "blabla..."
}
```

Reply:
fixed